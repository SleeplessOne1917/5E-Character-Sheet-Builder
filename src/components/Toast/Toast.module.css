.toast {
    padding: 1rem;
    display: grid;
    font-family: var(--font-display);
    color: white;
    grid-template-columns: 2fr 5fr 1fr;
    grid-template-rows: 1fr 1fr;
    grid-template-areas: "icon text x"
                        "icon text .";
    transform: translate(0, 100%);
    transition: transform 0.5s;
    pointer-events: all;
}

.open {
    transform: translateY(0) !important;
}

.toast-error {
    background-color: red;
}

.toast-success {
    background-color: green;
}

.toast-title {
    margin: 0;
    padding: 0;
    font-size: 1.8rem;
}

.toast-text {
    display: flex;
    flex-direction: column;
    justify-content: center;
    grid-area: text;
}

.icon {
    height: 3rem;
    width: 3rem;
}

.icon-container {
    grid-area: icon;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.x-icon {
    width: 2rem;
    height: 2rem;
    cursor: pointer;
}

.x-container {
    grid-area: x;
    display: flex;
    justify-content: flex-end;
    margin-top: 0.5rem;
    margin-right: 0.5rem;
}

.message {
    padding: 0;
    margin: 0;
    font-size: 1.1rem;
}

@media screen and (min-width: 768px) {
    .toast {
        bottom: 0;
        right: 0;
        position: absolute;
        width: 20rem;
        height: 9rem;
        transform: translate(100%, 0);
    }

    .open {
        transform: translateX(0) !important;
    }

    .toast-title {
        font-size: 2.3rem;
    }

    .message {
        font-size: 1.4rem;
    }
}

@media screen and (min-width: 1024px) {
    .toast {
        width: 25rem;
        height: 10rem;
    }

    .toast-title {
        font-size: 2.7rem;
    }

    .message {
        font-size: 1.6rem;
    }

    .icon {
        width: 4rem;
        height: 4rem;
    }
}

@media screen and (min-width: 1280px) {
    .toast {
        width: 30rem;
        height: 12rem;
    }

    .icon {
        width: 5rem;
        height: 5rem;
    }
}