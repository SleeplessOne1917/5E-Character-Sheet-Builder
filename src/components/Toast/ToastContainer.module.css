.toast-container {
    position: absolute;
    width: 100%;
    z-index: 10;
    pointer-events: none;
    height: 100%;
    top: 0;
    left: 0;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
}