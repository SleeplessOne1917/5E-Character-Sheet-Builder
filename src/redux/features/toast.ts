import { PayloadAction, createSlice } from '@reduxjs/toolkit';

import { ToastType } from '../../types/toast';

export type ToastState = {
	message: string;
	type: ToastType;
	isOpen: boolean;
	closeTimeoutSeconds: number;
};

export type ToastShowPayload = {
	message: string;
	type: ToastType;
	closeTimeoutSeconds: number;
};

const initialState: ToastState = {
	message: '',
	type: ToastType.success,
	isOpen: false,
	closeTimeoutSeconds: 0
};

const toastSlice = createSlice({
	name: 'toast',
	initialState,
	reducers: {
		show: (state, action: PayloadAction<ToastShowPayload>) => {
			return {
				isOpen: true,
				...action.payload
			};
		},
		hide: state => {
			return {
				...state,
				isOpen: false
			};
		}
	}
});

export const { show, hide } = toastSlice.actions;

export default toastSlice.reducer;
