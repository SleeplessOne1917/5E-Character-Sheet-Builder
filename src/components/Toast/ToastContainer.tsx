'use client';

import Toast from './Toast';
import classes from './ToastContainer.module.css';

const ToastContainer = (): JSX.Element => {
	return (
		<div className={classes['toast-container']}>
			<Toast />
		</div>
	);
};

export default ToastContainer;
