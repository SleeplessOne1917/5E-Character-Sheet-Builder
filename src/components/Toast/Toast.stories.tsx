import { ComponentMeta, ComponentStory } from '@storybook/react';
import { configureStore, createSlice } from '@reduxjs/toolkit';

import { Provider } from 'react-redux';
import Toast from './Toast';
import { ToastType } from '../../types/toast';

const MockStore = ({ mockState, children }) => (
	<Provider
		store={configureStore({
			reducer: {
				toast: createSlice({
					name: 'toast',
					initialState: {
						isOpen: true,
						closeTimeoutSeconds: 9999,
						...mockState
					},
					reducers: {}
				}).reducer
			}
		})}
	>
		{children}
	</Provider>
);

const ToastSb = {
	name: 'Components/Toast',
	component: Toast
} as ComponentMeta<typeof Toast>;

export default ToastSb;

const Template: ComponentStory<typeof Toast> = () => <Toast />;

export const Success = Template.bind({});
Success.decorators = [
	story => (
		<MockStore
			mockState={{
				type: ToastType.success,
				message: 'This is a success message.'
			}}
		>
			{story()}
		</MockStore>
	)
];

export const Error = Template.bind({});
Error.decorators = [
	story => (
		<MockStore
			mockState={{
				type: ToastType.error,
				message: 'This is an error message.'
			}}
		>
			{story()}
		</MockStore>
	)
];
