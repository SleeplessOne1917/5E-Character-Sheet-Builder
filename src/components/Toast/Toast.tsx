import {
	CheckCircleIcon,
	ExclaimationCircleIcon,
	XMarkIcon
} from '@heroicons/react/24/solid';
import { KeyboardEventHandler, useCallback, useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';

import { ToastType } from '../../types/toast';
import classes from './Toast.module.css';
import { cleanMessage } from '../../services/messageCleanerService';
import { handleKeyDownEvent } from '../../services/handlerService';
import { hide } from '../../redux/features/toast';

const Toast = (): JSX.Element => {
	const dispatch = useAppDispatch();
	const { message, closeTimeoutSeconds, isOpen, type } = useAppSelector(
		state => state.toast
	);

	useEffect(() => {
		let timer: NodeJS.Timeout;
		if (isOpen) {
			timer = setTimeout(hideToast, closeTimeoutSeconds * 1000);
		}
		return () => {
			if (timer) {
				clearTimeout(timer);
			}
		};
	}, [isOpen, closeTimeoutSeconds]); //eslint-disable-line

	const getClassName = useCallback(() => {
		return `${classes.toast} ${
			type === ToastType.error
				? classes['toast-error']
				: classes['toast-success']
		}${isOpen ? ` ${classes.open}` : ''}`;
	}, [isOpen, type]);

	const hideToast = useCallback(() => dispatch(hide()), [dispatch]);

	const hideToastKeyDown: KeyboardEventHandler<SVGSVGElement> = useCallback(
		event => {
			handleKeyDownEvent<SVGSVGElement>(event, hideToast);
		},
		[hideToast]
	);

	return (
		<div className={getClassName()} aria-label="toast" role="alert">
			<div className={classes['icon-container']}>
				{type === ToastType.error ? (
					<ExclaimationCircleIcon
						className={`${classes.icon} ${classes['error-icon']}`}
						role="img"
						aria-hidden="false"
						aria-label="Error"
					/>
				) : (
					<CheckCircleIcon
						className={`${classes.icon} ${classes['success-icon']}`}
						role="img"
						aria-hidden="false"
						aria-label="Success"
					/>
				)}
			</div>
			<div className={classes['toast-text']}>
				{type === ToastType.error ? (
					<div className={classes['toast-title']}>Error</div>
				) : (
					<div className={classes['toast-title']}>Success!</div>
				)}
				<p className={classes.message}>{cleanMessage(message)}</p>
			</div>
			<div className={classes['x-container']}>
				<XMarkIcon
					className={classes['x-icon']}
					onKeyDown={hideToastKeyDown}
					onClick={hideToast}
					tabIndex={isOpen ? 0 : -1}
					aria-hidden="false"
					aria-label="Close toast"
					role="button"
				/>
			</div>
		</div>
	);
};

export default Toast;
