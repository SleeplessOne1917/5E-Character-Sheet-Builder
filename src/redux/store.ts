import {
	AnyAction,
	ConfigureStoreOptions,
	ListenerEffect,
	Reducer,
	ThunkDispatch,
	configureStore,
	createListenerMiddleware
} from '@reduxjs/toolkit';
import editingCharacter, {
	EditingCharacterState
} from './features/editingCharacter';
import { get, set } from 'idb-keyval';
import { useEffect, useState } from 'react';

import generationMethod from './features/generationMethod';
import rollGroups from './features/rollGroups';
import toast from './features/toast';
import viewer from './features/viewer';

const getReducer = () => ({
	toast,
	viewer,
	editingCharacter: editingCharacter as Reducer<
		EditingCharacterState,
		AnyAction
	>,
	rollGroups,
	generationMethod
});

const indexedDbCacheListener = createListenerMiddleware();

const REDUX_CACHE_KEY = 'redux-cache';

const updateCache: ListenerEffect<
	AnyAction,
	unknown,
	ThunkDispatch<unknown, unknown, AnyAction>,
	unknown
> = (action, { getState }) => {
	set(REDUX_CACHE_KEY, getState());
};

const updateCachePredicate = () => true;

const getStore = async () => {
	indexedDbCacheListener.startListening({
		predicate: updateCachePredicate,
		effect: updateCache
	});

	const options: ConfigureStoreOptions = {
		reducer: getReducer(),
		middleware: getDefaultMiddleware =>
			getDefaultMiddleware().prepend(indexedDbCacheListener.middleware)
	};

	const reduxCache = await get(REDUX_CACHE_KEY);
	if (reduxCache) {
		options.preloadedState = reduxCache;
	}

	const theStore = configureStore(options);
	if (!reduxCache) {
		set(REDUX_CACHE_KEY, theStore.getState());
	}

	return theStore;
};

export const useStore = () => {
	const [theStore, setTheStore] = useState(getTestStore());
	const [loadingStore, setLoadingStore] = useState(true);

	useEffect(() => {
		getStore().then(s => {
			setTheStore(s);
			setLoadingStore(false);
		});

		return () => {
			indexedDbCacheListener.stopListening({
				predicate: updateCachePredicate,
				effect: updateCache
			});
		};
	}, [setTheStore]);

	return { store: theStore, loading: loadingStore };
};

export const getTestStore = (
	reducerOverrides?: Partial<ReturnType<typeof getReducer>>
) =>
	configureStore({
		reducer: { ...getReducer(), ...reducerOverrides },
		devTools: false
	});

const store = getTestStore();

export type RootState = ReturnType<typeof store.getState>;
export type Dispatch = typeof store.dispatch;
