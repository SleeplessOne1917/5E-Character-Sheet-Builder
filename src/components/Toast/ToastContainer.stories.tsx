import { ComponentMeta, ComponentStory } from '@storybook/react';

import Button from '../Button/Button';
import { Provider } from 'react-redux';
import ToastContainer from './ToastContainer';
import { ToastType } from '../../types/toast';
import { getTestStore } from '../../redux/store';
import { show } from '../../redux/features/toast';
import { useAppDispatch } from '../../hooks/reduxHooks';

export default {
	title: 'Components/ToastContainer',
	component: ToastContainer
} as ComponentMeta<typeof ToastContainer>;

const Template: ComponentStory<typeof ToastContainer> = (args: any) => (
	<ToastContainer {...args} />
);

export const Success = Template.bind({});
Success.decorators = [
	story => {
		const store = getTestStore();
		return (
			<Provider store={store}>
				<Button
					size="medium"
					onClick={() =>
						store.dispatch(
							show({
								closeTimeoutSeconds: 10,
								message: 'Success Message',
								type: ToastType.success
							})
						)
					}
				>
					Show Toast
				</Button>
				{story()}
			</Provider>
		);
	}
];

export const Error = Template.bind({});
Error.decorators = [
	story => {
		const store = getTestStore();
		return (
			<Provider store={store}>
				<Button
					size="medium"
					onClick={() =>
						store.dispatch(
							show({
								closeTimeoutSeconds: 10,
								message: 'Error Message',
								type: ToastType.error
							})
						)
					}
				>
					Show Toast
				</Button>
				{story()}
			</Provider>
		);
	}
];
