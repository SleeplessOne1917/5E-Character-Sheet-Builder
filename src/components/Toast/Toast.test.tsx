import '@testing-library/jest-dom';

import * as stories from './Toast.stories';

import { act, render, screen, waitFor } from '@testing-library/react';

import { Provider } from 'react-redux';
import Toast from './Toast';
import { ToastType } from '../../types/toast';
import { composeStories } from '@storybook/testing-react';
import { getStore } from '../../redux/store';
import { show } from '../../redux/features/toast';
import userEvent from '@testing-library/user-event';

const { Success, Error } = composeStories(stories);

it('renders correctly', () => {
	render(
		<Provider store={getStore()}>
			<Toast />
		</Provider>
	);

	expect(screen.getByRole('alert')).toMatchSnapshot();
});

it('success renders expected content', () => {
	render(<Success />);

	const toast = screen.getByRole('alert');
	expect(toast).toHaveClass('toast-success');
	expect(toast).not.toHaveClass('toast-error');
	expect(screen.getByRole('img')).not.toHaveClass('error-icon');
	expect(screen.getByRole('img')).toHaveClass('success-icon');
});

it('error renders expected content', () => {
	render(<Error />);

	const toast = screen.getByRole('alert');
	expect(toast).not.toHaveClass('toast-success');
	expect(toast).toHaveClass('toast-error');
	expect(screen.getByRole('img')).toHaveClass('error-icon');
	expect(screen.getByRole('img')).not.toHaveClass('success-icon');
});

it('does not contain open class when closed', () => {
	render(
		<Provider store={getStore()}>
			<Toast />
		</Provider>
	);

	expect(screen.getByRole('alert')).not.toHaveClass('open');
});

it('contains open class when open', () => {
	render(<Success />);

	expect(screen.getByRole('alert')).toHaveClass('open');
});

it('closes itself after closeTimeoutSeconds seconds', async () => {
	jest.useFakeTimers();
	const store = getStore();

	render(
		<Provider store={store}>
			<Toast />
		</Provider>
	);

	await waitFor(() =>
		store.dispatch(
			show({ type: ToastType.success, closeTimeoutSeconds: 3, message: 'foo' })
		)
	);

	expect(screen.getByRole('alert')).toHaveClass('open');

	act(() => {
		jest.advanceTimersByTime(3000);
	});
	expect(screen.getByRole('alert')).not.toHaveClass('open');
});

describe('x button tab index', () => {
	it('is -1 when closed', () => {
		render(
			<Provider store={getStore()}>
				<Toast />
			</Provider>
		);

		expect(screen.getByRole('button').tabIndex).toBe(-1);
	});

	it('is 0 when open', () => {
		render(<Success />);

		expect(screen.getByRole('button').tabIndex).toBe(0);
	});
});

describe('closes the toast when x button is pressed', () => {
	beforeEach(() => {
		jest.useFakeTimers();
		userEvent.setup({ delay: null });
	});

	it('with click', async () => {
		const store = getStore();

		render(
			<Provider store={store}>
				<Toast />
			</Provider>
		);

		await waitFor(() =>
			store.dispatch(
				show({
					type: ToastType.success,
					closeTimeoutSeconds: 999,
					message: 'foo'
				})
			)
		);

		expect(screen.getByRole('alert')).toHaveClass('open');

		await waitFor(async () => {
			await userEvent.click(screen.getByRole('button'));
		});

		expect(screen.getByRole('alert')).not.toHaveClass('open');
	});

	it('with enter', async () => {
		const store = getStore();

		render(
			<Provider store={store}>
				<Toast />
			</Provider>
		);

		await waitFor(() =>
			store.dispatch(
				show({
					type: ToastType.success,
					closeTimeoutSeconds: 999,
					message: 'foo'
				})
			)
		);

		expect(screen.getByRole('alert')).toHaveClass('open');

		await waitFor(async () => {
			await userEvent.tab();
			await userEvent.keyboard('{Enter}');
		});

		expect(screen.getByRole('alert')).not.toHaveClass('open');
	});

	it('with space', async () => {
		const store = getStore();

		render(
			<Provider store={store}>
				<Toast />
			</Provider>
		);

		await waitFor(() =>
			store.dispatch(
				show({
					type: ToastType.success,
					closeTimeoutSeconds: 999,
					message: 'foo'
				})
			)
		);

		expect(screen.getByRole('alert')).toHaveClass('open');

		await waitFor(async () => {
			await userEvent.tab();
			await userEvent.keyboard(' ');
		});

		expect(screen.getByRole('alert')).not.toHaveClass('open');
	});
});
