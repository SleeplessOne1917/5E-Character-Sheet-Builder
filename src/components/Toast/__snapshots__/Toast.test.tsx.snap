// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-label="toast"
  class="toast toast-success"
  role="alert"
>
  <div
    class="icon-container"
  >
    <svg
      aria-hidden="false"
      aria-label="Success"
      class="icon success-icon"
      fill="currentColor"
      role="img"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M2.25 12c0-5.385 4.365-9.75 9.75-9.75s9.75 4.365 9.75 9.75-4.365 9.75-9.75 9.75S2.25 17.385 2.25 12zm13.36-1.814a.75.75 0 10-1.22-.872l-3.236 4.53L9.53 12.22a.75.75 0 00-1.06 1.06l2.25 2.25a.75.75 0 001.14-.094l3.75-5.25z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <div
    class="toast-text"
  >
    <div
      class="toast-title"
    >
      Success!
    </div>
    <p
      class="message"
    />
  </div>
  <div
    class="x-container"
  >
    <svg
      aria-hidden="false"
      aria-label="Close toast"
      class="x-icon"
      fill="currentColor"
      role="button"
      tabindex="-1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;
